src/effects/backends/effectprocessor.h

```diff
@@ -16,23 +16,28 @@
 /// an EffectProcessorImpl subclass. Separating state from the DSP code allows
 /// memory allocation and deletion on the heap, which is slow, to be done on the
 /// main thread instead of potentially blocking the audio engine callback thread
-/// and causing audible glitches. EffectStates allocated on the main thread are
-/// passed as pointers to the EffectProcessorImpl in the audio callback thread
-/// via the EffectsMessenger.
+/// and causing audible glitches.
 ///
-/// Each EffectState instance is responsible for one routing of input signal to
-/// output signal. The base EffectProcessorImpl class handles the management
-/// of EffectStates. EffectProcessorImpl subclasses only need to be concerned
-/// with implementing the signal processing logic and providing metadata for
-/// describing the effect and its parameters with an EffectManifest.
+/// The base EffectState class does nothing on its own; each effect is responsible
+/// for subclassing it with whatever state it needs to maintain between cycles of
+/// the audio thread. The base EffectProcessorImpl class handles the management of
+/// EffectStates. EffectProcessorImpl subclasses only need to be concerned with
+/// implementing the signal processing logic and providing metadata for describing
+/// the effect and its parameters with an EffectManifest.
 ///
-/// Input signals can be any EngineChannel, but output channels are hardcoded in
-/// EngineMaster as the post-fader processing for the master mix and pre-fader
-/// processing for headphones. EffectStates are allocated when an input signal
-/// is enabled for a chain. Also, when a new effect is loaded to a chain,
-/// EffectStates are only allocated for input signals that are enabled at that
-/// time. This allows for scaling up to an arbitrary number of input signals
-/// without wasting a lot of memory.
+/// Each EffectState instance tracks the state for one combination of input signal
+/// and output signal. Input signals can be any EngineChannel, but output channels
+/// are hardcoded in EngineMaster as the post-fader processing for the main mix
+/// and pre-fader processing for headphones.
+///
+/// EffectStates allocated on the main thread are passed as pointers to the
+/// EffectProcessorImpl in the audio callback thread via the EffectsMessenger.
```

Comment:
I'm pretty confident that I understand now. So there is a single `effectProcessorImpl` class for doing the DSP for many different `EffectState`s, each `EffectState` representing some step in the processing pipeline.

Reply:
yes

Reply:
Could you add this summary here as well? I'm sorry to continue to ask this stuff, but a sentence like that would've helped me to understand this faster. It might seem redundant information, but in case of documentation I'd argue its beneficial since it helps the reader to reassure, they interpreted the information correctly.

Reply:
Sure, this is a really complex system so more documentation is better.

Reply:
How is this? c5d522d2afa355c73c010a0d3f3bec29bf6fc2d7

Reply:
I don't like "channels" because you can mistake it for literal audio channels (Left/Right) and different effectstates can still act on the same mixer channel at different parts of the pipeline. 
How about this: 

There can be many `EffectState` instances for one `EffectProcessorImpl` instance, allowing a single `EffectProcessorImpl` to maintain an independent `EffectState` for each occurrence in mixxx audio processing pipeline. So `EffectState`s  can be seen as a way to provide context to the DSP code inside `EffectProcessorImpl`, thus `EffectProcessorImpl` should not store any context information / state on their own.

Reply:
IMO that implies that one EffectProcessorImpl instance could be at different points in the signal path, which does not happen.

Reply:
Good to know, because then I misunderstood, so to which scope is an EffectProcessorImpl instance limited to? I guess a chain?

Reply:
There is a 1:1 relationship between EngineEffect and EffectProcessor as noted in the EngineEffect documentation.

Reply:
I'm not sure we have the same understanding of "step in the processing pipeline". Multiple input signals can go through the same EffectProcessor (simultaneously) but that EffectProcessor instance is at the same point in each input signal's processing.

Reply:
Sorry, I got confused. That was an oversimplification. For the QuickEffectChains and StandardEffectChains, one EffectProcessor instance processes the same input signal at different points in the signal flow, once postfader for the main mix and once prefader for headphones.

Reply:
Right, so each EffectChain contains only one EffectProcessor instance per effect. That EffectProcessor can be re-used multiple times in a chain with different EffectStates? This seems overcomplicated to me...

Reply:
It is as complicated as it needs to be. I thought the documentation in effectsprocessor.h explained why this complexity is required. Please reread it and comment if you don't understand.

Reply:
Well I understand why this separation is a good idea. At this point, I understand this architecture enough so I can start questioning it. My problem now is likely that my mental model of how audio is routed within mixxx is probably not correct, but thats not really the fault of the effectssystem.

Reply:
There was months of extensive discussion about this in #1254 if you want to review the reasoning for how it got where it is.

Reply:
High-level overviews like this would be very useful to document in the entrypoint of a system

```quote
New signal flow:

pregain ---> fader and crossfader attenuation --> channel effect processing  --> mix channels into master mix and crossfader busses --> master effect processing
         \
           --> channel effect processing --> effect chain dry/wet knob --> mix channels into headphone mix --> headphone effect processing
```

Reply:
Sure, I could put that in EngineEffectsManager.

Reply:
Wherever it makes sense and its seen by people intending to make changes to that. Thank you.

Reply:
d836e80743701f8132bd185c8a8af7324b4b7ff3

Reply:
unfortately clang tidy (and even gcc?) goes absolutely crazy over this change....

> multi-line comment [-Werror=comment]

Reply:
Yes I fixed that and force pushed: fb905f8

I replaced the `\` with `|`